### PR TITLE
leverage new GeoPointInPolygon query for WITHIN

### PR DESCRIPTION
**Pull request by chaudum**

### CODE REVIEW

File: sql/src/main/java/io/crate/lucene/LuceneQueryBuilder.java [hunk 1]
```diff
@@ -829,13 +833,35 @@ private Query getQuery(Function inner, Context context) {
...
+                            coordinates = Arrays.copyOf(coordinates, coordinates.length + 1);
+                            coordinates[coordinates.length-1] = coordinates[0];
+                        }
+                        final double[] lats = new double[coordinates.length];
+                        final double[] lons = new double[coordinates.length];
+                        for (int i = 0; i < coordinates.length; i++) {
+                            lats[i] = coordinates[i].y;
+                            lons[i] = coordinates[i].x;
+                        }
+                        // if index created V_2_2 use (soon to be legacy) numeric encoding postings format
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/lucene/LuceneQueryBuilder.java]: We've never had ES 2.2 released, so Either it's < 2.2 or >= 2.3

File: sql/src/main/java/io/crate/lucene/LuceneQueryBuilder.java [hunk 2]
```diff
@@ -829,13 +833,35 @@ private Query getQuery(Function inner, Context context) {
...
+                    final Version indexCreated = Version.indexCreated(context.indexCache.indexSettings());
                     Coordinate[] coordinates = geometry.getCoordinates();
-                    GeoPoint[] points = new GeoPoint[coordinates.length];
-                    for (int i = 0; i < coordinates.length; i++) {
-                        Coordinate coordinate = coordinates[i];
-                        points[i] = new GeoPoint(coordinate.y, coordinate.x);
+                    final Query query;
+                    if (indexCreated.before(Version.V_2_2_0)) {
+                        GeoPoint[] points = new GeoPoint[coordinates.length];
+                        for (int i = 0; i < coordinates.length; i++) {
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/lucene/LuceneQueryBuilder.java]: can you extract some of the logic here into separate method? getQuery is already rather large.

File: sql/src/main/java/io/crate/lucene/LuceneQueryBuilder.java [hunk 3]
```diff
@@ -822,21 +826,50 @@ private Query getQuery(Function inner, Context context) {
...
+                    final double[] lats = new double[coordinates.length];
+                    final double[] lons = new double[coordinates.length];
+                    for (int i = 0; i < coordinates.length; i++) {
+                        lats[i] = coordinates[i].y;
+                        lons[i] = coordinates[i].x;
+                    }
+                    // if index was created before V_2_3 -> use NUMERIC encoded postings format, otherwise PREFIX
+                    final GeoPointField.TermEncoding encoding = indexCreated.before(Version.V_2_3_0) ?
+                        GeoPointField.TermEncoding.NUMERIC : GeoPointField.TermEncoding.PREFIX;
+                    query = new GeoPointInPolygonQuery(fieldData.getFieldNames().indexName(), encoding, lons, lats);
```
Comment (chaudum) [inline sql/src/main/java/io/crate/lucene/LuceneQueryBuilder.java]: @mfussenegger 
should we use?
```java
if (indexCreated.before(Version.V_2_3_0)) {
    ...
} else  {
    query = new GeoPointInPolygonQuery(fieldData.getFieldNames().indexName(), GeoPointField.TermEncoding.PREFIX, lons, lats);
}
```

**Comment by mfussenegger** [comment]
lgtm

**Comment by bogensberger** [comment]
lgtm

**Pull request merged**